File: api_client.js
/**
 * REST API Client Examples
 * Comprehensive examples for sending POST and GET requests using different JavaScript approaches.
 */

// Using fetch (modern browsers and Node.js 18+)
class APIClient {
    /**
     * A comprehensive REST API client class with examples for GET and POST requests.
     * 
     * This class demonstrates different approaches to making HTTP requests:
     * - Using fetch API (modern)
     * - Using XMLHttpRequest (legacy)
     * - Error handling and response processing
     */
    
    constructor(baseURL = 'https://api.example.com', timeout = 30000) {
        this.baseURL = baseURL.replace(/\/$/, '');
        this.timeout = timeout;
        
        // Common headers
        this.defaultHeaders = {
            'Content-Type': 'application/json',
            'User-Agent': 'APIClient/1.0'
        };
    }
    
    /**
     * Send a GET request using fetch API
     * @param {string} endpoint - API endpoint (e.g., '/users')
     * @param {Object} params - Query parameters
     * @param {Object} headers - Additional headers
     * @returns {Promise<Object>} Response data
     */
    async getRequest(endpoint, params = {}, headers = {}) {
        const url = new URL(`${this.baseURL}${endpoint}`);
        
        // Add query parameters
        Object.keys(params).forEach(key => {
            url.searchParams.append(key, params[key]);
        });
        
        const requestHeaders = { ...this.defaultHeaders, ...headers };
        
        try {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), this.timeout);
            
            const response = await fetch(url.toString(), {
                method: 'GET',
                headers: requestHeaders,
                signal: controller.signal
            });
            
            clearTimeout(timeoutId);
            
            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }
            
            return await response.json();
            
        } catch (error) {
            console.error('GET request failed:', error);
            throw error;
        }
    }
    
    /**
     * Send a POST request using fetch API
     * @param {string} endpoint - API endpoint (e.g., '/users')
     * @param {Object} data - Data to send in the request body
     * @param {Object} headers - Additional headers
     * @returns {Promise<Object>} Response data
     */
    async postRequest(endpoint, data = {}, headers = {}) {
        const url = `${this.baseURL}${endpoint}`;
        const requestHeaders = { ...this.defaultHeaders, ...headers };
        
        try {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), this.timeout);
            
            const response = await fetch(url, {
                method: 'POST',
                headers: requestHeaders,
                body: JSON.stringify(data),
                signal: controller.signal
            });
            
            clearTimeout(timeoutId);
            
            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }
            
            return await response.json();
            
        } catch (error) {
            console.error('POST request failed:', error);
            throw error;
        }
    }
    
    /**
     * Send a GET request using XMLHttpRequest (legacy approach)
     * @param {string} endpoint - API endpoint
     * @param {Object} params - Query parameters
     * @returns {Promise<Object>} Response data
     */
    getRequestXHR(endpoint, params = {}) {
        return new Promise((resolve, reject) => {
            const url = new URL(`${this.baseURL}${endpoint}`);
            
            // Add query parameters
            Object.keys(params).forEach(key => {
                url.searchParams.append(key, params[key]);
            });
            
            const xhr = new XMLHttpRequest();
            
            xhr.open('GET', url.toString(), true);
            xhr.setRequestHeader('Content-Type', 'application/json');
            
            xhr.timeout = this.timeout;
            
            xhr.onload = function() {
                if (xhr.status >= 200 && xhr.status < 300) {
                    try {
                        const data = JSON.parse(xhr.responseText);
                        resolve(data);
                    } catch (error) {
                        reject(new Error('Invalid JSON response'));
                    }
                } else {
                    reject(new Error(`HTTP error! status: ${xhr.status}`));
                }
            };
            
            xhr.onerror = function() {
                reject(new Error('Network error'));
            };
            
            xhr.ontimeout = function() {
                reject(new Error('Request timeout'));
            };
            
            xhr.send();
        });
    }
    
    /**
     * Send a POST request using XMLHttpRequest (legacy approach)
     * @param {string} endpoint - API endpoint
     * @param {Object} data - Data to send
     * @returns {Promise<Object>} Response data
     */
    postRequestXHR(endpoint, data = {}) {
        return new Promise((resolve, reject) => {
            const url = `${this.baseURL}${endpoint}`;
            const xhr = new XMLHttpRequest();
            
            xhr.open('POST', url, true);
            xhr.setRequestHeader('Content-Type', 'application/json');
            
            xhr.timeout = this.timeout;
            
            xhr.onload = function() {
                if (xhr.status >= 200 && xhr.status < 300) {
                    try {
                        const responseData = JSON.parse(xhr.responseText);
                        resolve(responseData);
                    } catch (error) {
                        reject(new Error('Invalid JSON response'));
                    }
                } else {
                    reject(new Error(`HTTP error! status: ${xhr.status}`));
                }
            };
            
            xhr.onerror = function() {
                reject(new Error('Network error'));
            };
            
            xhr.ontimeout = function() {
                reject(new Error('Request timeout'));
            };
            
            xhr.send(JSON.stringify(data));
        });
    }
    
    /**
     * Upload file using FormData
     * @param {string} endpoint - API endpoint
     * @param {File} file - File to upload
     * @param {Object} additionalData - Additional form data
     * @returns {Promise<Object>} Response data
     */
    async uploadFile(endpoint, file, additionalData = {}) {
        const url = `${this.baseURL}${endpoint}`;
        const formData = new FormData();
        
        formData.append('file', file);
        
        // Add additional data
        Object.keys(additionalData).forEach(key => {
            formData.append(key, additionalData[key]);
        });
        
        try {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), this.timeout);
            
            const response = await fetch(url, {
                method: 'POST',
                body: formData,
                signal: controller.signal
                // Note: Don't set Content-Type header for FormData
            });
            
            clearTimeout(timeoutId);
            
            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }
            
            return await response.json();
            
        } catch (error) {
            console.error('File upload failed:', error);
            throw error;
        }
    }
}

// Example usage functions
async function exampleGetRequests() {
    console.log('=== GET Request Examples ===');
    
    const client = new APIClient('https://jsonplaceholder.typicode.com');
    
    try {
        // Simple GET request
        const users = await client.getRequest('/users');
        console.log(`Retrieved ${users.length} users`);
        
        // GET request with parameters
        const posts = await client.getRequest('/posts', { userId: 1 });
        console.log(`Retrieved ${posts.length} posts for user 1`);
        
    } catch (error) {
        console.error('Error:', error.message);
    }
}

async function examplePostRequests() {
    console.log('\n=== POST Request Examples ===');
    
    const client = new APIClient('https://jsonplaceholder.typicode.com');
    
    const newPost = {
        title: 'Test Post',
        body: 'This is a test post',
        userId: 1
    };
    
    try {
        const response = await client.postRequest('/posts', newPost);
        console.log(`Created post with ID: ${response.id}`);
        
    } catch (error) {
        console.error('Error:', error.message);
    }
}

async function exampleWithAuthentication() {
    console.log('\n=== Authentication Example ===');
    
    const client = new APIClient('https://api.example.com');
    
    const headers = {
        'Authorization': 'Bearer your_token_here',
        'X-API-Key': 'your_api_key_here'
    };
    
    try {
        const data = await client.getRequest('/protected-endpoint', {}, headers);
        console.log('Authenticated request successful');
        
    } catch (error) {
        console.error('Error:', error.message);
    }
}

async function exampleFileUpload() {
    console.log('\n=== File Upload Example ===');
    
    const client = new APIClient('https://api.example.com');
    
    // In a real browser environment, you would get the file from an input element
    // const fileInput = document.getElementById('fileInput');
    // const file = fileInput.files[0];
    
    // For demonstration, we'll show the structure
    console.log('File upload structure (requires actual file in browser environment):');
    console.log('const file = fileInput.files[0];');
    console.log('await client.uploadFile("/upload", file, { description: "Uploaded file" });');
}

// Node.js specific examples (if running in Node.js environment)
async function nodeJSExamples() {
    console.log('\n=== Node.js Specific Examples ===');
    
    // Using node-fetch (if installed)
    try {
        const fetch = (await import('node-fetch')).default;
        
        const response = await fetch('https://jsonplaceholder.typicode.com/users');
        const users = await response.json();
        console.log(`Node.js fetch: Retrieved ${users.length} users`);
        
    } catch (error) {
        console.log('node-fetch not available or error occurred');
    }
    
    // Using axios (if installed)
    try {
        const axios = (await import('axios')).default;
        
        const response = await axios.get('https://jsonplaceholder.typicode.com/users');
        console.log(`Axios: Retrieved ${response.data.length} users`);
        
    } catch (error) {
        console.log('axios not available or error occurred');
    }
}

// Run examples
async function runExamples() {
    console.log('REST API Client Examples\n');
    
    await exampleGetRequests();
    await examplePostRequests();
    await exampleWithAuthentication();
    await exampleFileUpload();
    
    // Check if we're in Node.js environment
    if (typeof window === 'undefined') {
        await nodeJSExamples();
    }
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { APIClient, runExamples };
}

// Run examples if this file is executed directly
if (typeof window !== 'undefined' || require.main === module) {
    runExamples().catch(console.error);
}

File: api_client.py
#!/usr/bin/env python3
"""
REST API Client Examples
Comprehensive examples for sending POST and GET requests using different Python libraries.
"""

import requests
import json
import urllib.request
import urllib.parse
from typing import Dict, Any, Optional


class APIClient:
    """
    A comprehensive REST API client class with examples for GET and POST requests.
    
    This class demonstrates different approaches to making HTTP requests:
    - Using requests library (recommended)
    - Using urllib (built-in)
    - Error handling and response processing
    """
    
    def __init__(self, base_url: str = "https://api.example.com", timeout: int = 30):
        """
        Initialize the API client.
        
        Args:
            base_url (str): Base URL for the API
            timeout (int): Request timeout in seconds
        """
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.session = requests.Session()
        
        # Common headers
        self.default_headers = {
            'Content-Type': 'application/json',
            'User-Agent': 'APIClient/1.0'
        }
    
    def get_request_requests(self, endpoint: str, params: Optional[Dict] = None, 
                           headers: Optional[Dict] = None) -> Dict[str, Any]:
        """
        Send a GET request using the requests library.
        
        Args:
            endpoint (str): API endpoint (e.g., '/users')
            params (dict, optional): Query parameters
            headers (dict, optional): Additional headers
            
        Returns:
            dict: Response data
            
        Raises:
            requests.RequestException: If the request fails
        """
        url = f"{self.base_url}{endpoint}"
        request_headers = {**self.default_headers, **(headers or {})}
        
        try:
            response = self.session.get(
                url,
                params=params,
                headers=request_headers,
                timeout=self.timeout
            )
            response.raise_for_status()
            return response.json()
            
        except requests.exceptions.RequestException as e:
            print(f"GET request failed: {e}")
            raise
    
    def post_request_requests(self, endpoint: str, data: Dict[str, Any], 
                            headers: Optional[Dict] = None) -> Dict[str, Any]:
        """
        Send a POST request using the requests library.
        
        Args:
            endpoint (str): API endpoint (e.g., '/users')
            data (dict): Data to send in the request body
            headers (dict, optional): Additional headers
            
        Returns:
            dict: Response data
            
        Raises:
            requests.RequestException: If the request fails
        """
        url = f"{self.base_url}{endpoint}"
        request_headers = {**self.default_headers, **(headers or {})}
        
        try:
            response = self.session.post(
                url,
                json=data,
                headers=request_headers,
                timeout=self.timeout
            )
            response.raise_for_status()
            return response.json()
            
        except requests.exceptions.RequestException as e:
            print(f"POST request failed: {e}")
            raise
    
    def get_request_urllib(self, endpoint: str, params: Optional[Dict] = None) -> Dict[str, Any]:
        """
        Send a GET request using urllib (built-in library).
        
        Args:
            endpoint (str): API endpoint
            params (dict, optional): Query parameters
            
        Returns:
            dict: Response data
        """
        url = f"{self.base_url}{endpoint}"
        
        if params:
            query_string = urllib.parse.urlencode(params)
            url = f"{url}?{query_string}"
        
        try:
            with urllib.request.urlopen(url, timeout=self.timeout) as response:
                data = response.read()
                return json.loads(data.decode('utf-8'))
                
        except Exception as e:
            print(f"GET request failed: {e}")
            raise
    
    def post_request_urllib(self, endpoint: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Send a POST request using urllib (built-in library).
        
        Args:
            endpoint (str): API endpoint
            data (dict): Data to send
            
        Returns:
            dict: Response data
        """
        url = f"{self.base_url}{endpoint}"
        json_data = json.dumps(data).encode('utf-8')
        
        request = urllib.request.Request(
            url,
            data=json_data,
            headers={'Content-Type': 'application/json'},
            method='POST'
        )
        
        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                data = response.read()
                return json.loads(data.decode('utf-8'))
                
        except Exception as e:
            print(f"POST request failed: {e}")
            raise


# Example usage functions
def example_get_requests():
    """Example of using requests library for GET requests."""
    client = APIClient("https://jsonplaceholder.typicode.com")
    
    # Simple GET request
    try:
        users = client.get_request_requests('/users')
        print(f"Retrieved {len(users)} users")
        
        # GET request with parameters
        posts = client.get_request_requests('/posts', params={'userId': 1})
        print(f"Retrieved {len(posts)} posts for user 1")
        
    except Exception as e:
        print(f"Error: {e}")


def example_post_requests():
    """Example of using requests library for POST requests."""
    client = APIClient("https://jsonplaceholder.typicode.com")
    
    # POST request with data
    new_post = {
        "title": "Test Post",
        "body": "This is a test post",
        "userId": 1
    }
    
    try:
        response = client.post_request_requests('/posts', new_post)
        print(f"Created post with ID: {response.get('id')}")
        
    except Exception as e:
        print(f"Error: {e}")


def example_with_authentication():
    """Example with authentication headers."""
    client = APIClient("https://api.example.com")
    
    # Add authentication header
    headers = {
        'Authorization': 'Bearer your_token_here',
        'X-API-Key': 'your_api_key_here'
    }
    
    try:
        # Authenticated GET request
        data = client.get_request_requests('/protected-endpoint', headers=headers)
        print("Authenticated request successful")
        
    except Exception as e:
        print(f"Error: {e}")


def example_file_upload():
    """Example of uploading files with POST request."""
    client = APIClient("https://api.example.com")
    
    # For file uploads, you would use files parameter
    files = {
        'file': ('filename.txt', open('filename.txt', 'rb'), 'text/plain')
    }
    
    data = {
        'description': 'Uploaded file'
    }
    
    try:
        response = client.session.post(
            f"{client.base_url}/upload",
            files=files,
            data=data,
            timeout=client.timeout
        )
        response.raise_for_status()
        print("File uploaded successfully")
        
    except Exception as e:
        print(f"Upload failed: {e}")


if __name__ == "__main__":
    print("=== REST API Client Examples ===\n")
    
    print("1. GET Request Example:")
    example_get_requests()
    print()
    
    print("2. POST Request Example:")
    example_post_requests()
    print()
    
    print("3. Authentication Example:")
    example_with_authentication()
    print()
    
    print("4. File Upload Example:")
    example_file_upload()